### PR TITLE
fix(rust): don't mark pure-annotated local calls as execution-order-sensitive

### PR DESCRIPTION
**Pull request by Dunqing**
Pure-annotated calls to local functions (e.g. `/*#__PURE__*/ foo()`) should not trigger `ExecutionOrderSensitive`, since the annotation is a developer contract that the call is side-effect-free. Only `Unknown` and `GlobalVarAccess` flags should trigger wrapping.

This also removes the now-unused `SideEffectDetail::PureAnnotation` flag.

### CODE REVIEW

File: crates/rolldown/tests/rolldown/function/experimental/strict_execution_order/issue_4920/artifacts.snap [hunk 1]
```diff
@@ -38,13 +38,9 @@ import nodeAssert from "node:assert";
...
 }
-var dep_default;
-var init_dep = __esmMin((() => {
-	dep_default = /* @__PURE__ */ foo();
-}));
+var dep_default = /* @__PURE__ */ foo();
 //#endregion
 __esmMin((() => {
-	init_dep();
 	nodeAssert.strictEqual(globalThis.value, 1);
```
Comment (Dunqing) [inline crates/rolldown/tests/rolldown/function/experimental/strict_execution_order/issue_4920/artifacts.snap:44]: The diff is back to the state before https://github.com/rolldown/rolldown/pull/5498

**Comment by IWANABETHATGUY** [comment]
https://github.com/rolldown/rolldown/issues/5303

**Comment by Dunqing** [comment]
> #5303

This change doesn't affect #5303, and `GlobalVarAccess` actually already covered this case.

File: crates/rolldown/tests/rolldown/function/experimental/strict_execution_order/issue_4920/artifacts.snap (hunk 1 above)
Comment (IWANABETHATGUY) [reply crates/rolldown/tests/rolldown/function/experimental/strict_execution_order/issue_4920/artifacts.snap:44]: Since `foo` mutate the globalValue, it should be wrapped to ensure the correct execution order

**Comment by IWANABETHATGUY** [comment]
Here is an example worked on the main branch and not in pr branch 
https://repl.rolldown.rs/#eNptks1OwzAQhF/FMoeWKjgt4hTEAUGPCIT4uUSKTLIpRs46OHapiPLu2E5LDe0tmd3Znf3kntY06ykvWquh0MAr0Oyj8xrS7FBPaOnk2mJphEKyAnOnKpiekj5HQjQYq50q1RuXT++iY51T2hcuLVzmOOSYI2xapQ2poOZWGpLOTori4flxWRSzdD/PddMhoVpJWakvZKXCWqyiZEcqYzbRhPm93yAQbkKZDKTWqiGTnW3iFhxkiR3TcJDA1pqMjDNZA4azICW+qKwJ1dBJSGe0KM1yA6X1bO61Q5YRoy34+hAsbiFo0QAaLn+NCm+h4Vi9at62AleRyTHbofguAssIQaSMpx/lTq7Ixfl2RsMFRgN2v3/AoeN/3XXgPkdkXsh4UDy1bduEpfv9kb4OO0cnS/+/nxH7fgXbQvu0XE6DNXFpw8kJBZp5EIMLvnb5FmzO5me6ZIs5HX4Aj5fw+w==

File: crates/rolldown/tests/rolldown/function/experimental/strict_execution_order/issue_4920/artifacts.snap (hunk 1 above)
Comment (Dunqing) [reply crates/rolldown/tests/rolldown/function/experimental/strict_execution_order/issue_4920/artifacts.snap:44]: Thank you for reviewing. I indeed found a case where the mixed issue-4920 and issue-5303 can be broken after this change. Sorry for wasting your time.

**Comment by Dunqing** [comment]
> Here is an example worked on the main branch and not in pr branch https://repl.rolldown.rs/#eNptks1OwzAQhF/FMoeWKjgt4hTEAUGPCIT4uUSKTLIpRs46OHapiPLu2E5LDe0tmd3Znf3kntY06ykvWquh0MAr0Oyj8xrS7FBPaOnk2mJphEKyAnOnKpiekj5HQjQYq50q1RuXT++iY51T2hcuLVzmOOSYI2xapQ2poOZWGpLOTori4flxWRSzdD/PddMhoVpJWakvZKXCWqyiZEcqYzbRhPm93yAQbkKZDKTWqiGTnW3iFhxkiR3TcJDA1pqMjDNZA4azICW+qKwJ1dBJSGe0KM1yA6X1bO61Q5YRoy34+hAsbiFo0QAaLn+NCm+h4Vi9at62AleRyTHbofguAssIQaSMpx/lTq7Ixfl2RsMFRgN2v3/AoeN/3XXgPkdkXsh4UDy1bduEpfv9kb4OO0cnS/+/nxH7fgXbQvu0XE6DNXFpw8kJBZp5EIMLvnb5FmzO5me6ZIs5HX4Aj5fw+w==

Oh, I missed this. Yes you are right. https://github.com/rolldown/rolldown/pull/8909#discussion_r2992429014

File: crates/rolldown/tests/rolldown/function/experimental/strict_execution_order/issue_4920/artifacts.snap (hunk 1 above)
Comment (IWANABETHATGUY) [reply crates/rolldown/tests/rolldown/function/experimental/strict_execution_order/issue_4920/artifacts.snap:44]: > Thank you for reviewing. I indeed found a case where the mixed issue-4920 and issue-5303 can be broken after this change. Sorry for wasting your time.

Don't sweat it! It’s definitely not a waste of time—it was a tricky edge case. I'm just happy to help.

Comment (Dunqing) [reply crates/rolldown/tests/rolldown/function/experimental/strict_execution_order/issue_4920/artifacts.snap:44]: Do you think it is worth adding a test?

Comment (IWANABETHATGUY) [reply crates/rolldown/tests/rolldown/function/experimental/strict_execution_order/issue_4920/artifacts.snap:44]: Yeah, let's do that.
